File: specsy-core/src/main/scala/org/specsy/core/SpecClassRunner.scala
// Copyright © 2010-2012, Esko Luontola <www.orfjackal.net>
// This software is released under the Apache License 2.0.
// The license text is at http://www.apache.org/licenses/LICENSE-2.0

package org.specsy.core

import java.lang.reflect.InvocationTargetException
import java.util.concurrent.Executor
import fi.jumi.api.drivers.SuiteNotifier

class SpecClassRunner(testClass: Class[_], notifier: SuiteNotifier, executor: Executor) extends Runnable {
  def run() {
    runSpec(c => {
      c.bootstrap(testClass.getSimpleName, new Closure {
        def run() {
          ContextDealer.prepare(c)
          try {
            testClass.getConstructor().newInstance()
          } catch {
            case e: InvocationTargetException => throw e.getTargetException
          }
        }
      })
    })
  }

  private def runSpec(spec: Context => Unit) {
    val runner = new SpecRun(spec, notifier, executor)
    runner.run()
  }
}

File: specsy-core/src/main/java/org/specsy/core/SpecClassRunner.java
// Copyright © 2010-2012, Esko Luontola <www.orfjackal.net>
// This software is released under the Apache License 2.0.
// The license text is at http://www.apache.org/licenses/LICENSE-2.0

package org.specsy.core;

import fi.jumi.api.drivers.SuiteNotifier;

import java.lang.reflect.InvocationTargetException;
import java.util.concurrent.Executor;

public class SpecClassRunner implements Runnable {

    private final Class<?> testClass;
    private final SuiteNotifier notifier;
    private final Executor executor;

    public SpecClassRunner(Class<?> testClass, SuiteNotifier notifier, Executor executor) {
        this.testClass = testClass;
        this.notifier = notifier;
        this.executor = executor;
    }

    @Override
    public void run() {
        new SpecRun(new TestClassBootstrap(), notifier, executor).run();
    }

    private class TestClassBootstrap implements Spec {
        @Override
        public void run(final Context context) {
            context.bootstrap(testClass.getSimpleName(), new Closure() {
                @Override
                public void run() throws Throwable {
                    ContextDealer.prepare(context);
                    try {
                        testClass.getConstructor().newInstance();
                    } catch (InvocationTargetException e) {
                        throw e.getTargetException();
                    }
                }
            });
        }
    }

}
